pkg/client/ssa_client.go

```diff
@@ -0,0 +1,218 @@
+	var ua string
+	if cfg != nil {
+		ua = cfg.UserAgent
+	}
+
+	if len(ua) == 0 {
+		ua = rest.DefaultKubernetesUserAgent()
+	}
+
+	name := strings.Split(ua, "/")[0]
+	return name
+}
+
+type ssaApplyObjectConfiguration struct {
+	owner           metav1.Object
+	newLabels       map[string]string
+	determineUpdate bool
+	skipIf          func(client.Object) bool
+}
+
+func newSsaApplyObjectConfiguration(options ...SsaApplyObjectOption) ssaApplyObjectConfiguration {
+	config := ssaApplyObjectConfiguration{}
+	for _, apply := range options {
+		apply(&config)
+	}
+	return config
+}
+
+// SsaApplyObjectOption an option when creating or updating a resource
+type SsaApplyObjectOption func(*ssaApplyObjectConfiguration)
+
+// SsaSetOwner sets the owner of the resource (default: `nil`)
+func SsaSetOwner(owner metav1.Object) SsaApplyObjectOption {
+	return func(config *ssaApplyObjectConfiguration) {
+		config.owner = owner
+	}
+}
+
+// SkipIf will cause the apply function skip the update of the object if
+// the provided function returns true. The supplied object is guaranteed to
+// have its GVK set.
+func SkipIf(test func(client.Object) bool) SsaApplyObjectOption {
+	return func(config *ssaApplyObjectConfiguration) {
+		config.skipIf = test
+	}
+}
+
+// DetermineUpdate instructs the ApplyObject function to truly test whether the object
+// was updated or not during the apply operation. By default, the function always returns
+// true on success which is more efficient and sufficient in most circumstances.
+func DetermineUpdate(value bool) SsaApplyObjectOption {
+	return func(config *ssaApplyObjectConfiguration) {
+		config.determineUpdate = value
+	}
+}
+
+// EnsureLabels makes sure that the provided labels are applied to the object even if
+// the supplied object doesn't have them set.
+func EnsureLabels(labels map[string]string) SsaApplyObjectOption {
+	return func(config *ssaApplyObjectConfiguration) {
+		config.newLabels = labels
+	}
+}
+
+// Configure sets the owner reference and merges the labels. Other options modify the logic
+// of apply function and therefore need to be checked manually.
+func (c *ssaApplyObjectConfiguration) Configure(obj client.Object, s *runtime.Scheme) error {
+	if c.owner != nil {
+		if err := controllerutil.SetControllerReference(c.owner, obj, s); err != nil {
+			return err
+		}
+	}
+	MergeLabels(obj, c.newLabels)
+
+	return nil
+}
+
+// ApplyObject creates the object if is missing and if the owner object is provided, then it's set as a controller reference.
+// If the objects exists then when the spec content has changed (based on the content of the annotation in the original object) then it
+// is automatically updated. If it looks to be same then based on the value of forceUpdate param it updates the object or not.
+// The return boolean says if the object was either created or updated (`true`). If nothing changed (ie, the generation was not
+// incremented by the server), then it returns `false`.
+//
```

Comment:
this looks like a copy-paste of the already existing ApplyObject method

Reply:
Indeed, I will reword it.